cmake/Googletest.cmake

```diff
@@ -79,4 +85,10 @@ macro(fetch_googletest)
         set(GTEST_MAIN_LIBRARIES ${GTEST_MAIN_LIBRARY})
     endif()
 
+    set(GTEST_LIBS ${GTEST_LIBRARIES})
+    if(${CMAKE_SYSTEM_NAME} MATCHES "Windows" AND NOT CMAKE_BUILD_TYPE MATCHES Debug)
+        set(GTEST_LIBS "${GTEST_ROOT}/bin/gtest.dll")
```

Comment:
Good catch! There is actually a bug here on debug!!! 
I changed the logics a bit and added comments so it is more clear. See 26e3834 for fix.

Reply:
Looking at the latest changes... and I'm still a bit confused.  Isn't the code branch above ("Setting GTest libraries with debug..."), already setting `GTEST_LIBRARIES` to be `${GTEST_ROOT}/bin/gtestd.dll`, if we're on Windows + Debug?  I think the only situation where it wouldn't be would be if it were "found" by the `find_package` (ie, if `GTEST_LIBRARY_DEBUG` did not match `GTEST_LIBRARY_DEBUG-NOTFOUND`).

It just seems like we're fixing the same bug twice, in two separate places.

Reply:
>  Isn't the code branch above ("Setting GTest libraries with debug..."), already setting GTEST_LIBRARIES to be ${GTEST_ROOT}/bin/gtestd.dll, if we're on Windows + Debug? 

Correct, we explicitly set `gtest_library` to point at `bin/gtestd.dll` in Debug. However, if it is release, ${GTEST_LIBRARIES} returns `lib/gtest.lib` which is obviously what we don't want.

I see the redundancy if some of the logics. I will come back to it later.

Reply:
Hmm... so, in general, on windows, is the `find_package` returning `.lib` entries in `GTEST_LIBRARIES`, or `.dll`?

Reply:
@elrond79 Sadly it returns `.lib` :) 

That's why I hate CMake modules that are out there. They are either incomplete, buggy, or it is a copy of someone's Module with different names, rules,etc....

I think we should just right our own Find modules here but that's for another day :)

Reply:
Actually... I though the .lib files are what you need for linking when building on windows - ie: 

https://stackoverflow.com/questions/7845886/linking-dll-in-visual-studio

I know relatively little about windows building, though, so... maybe that's not an absolute? Are there some circumstances where you need the .dll?

In any case, I think we need consistent behavior - right now, in some code paths GTEST_LIBRARIES will end up with a .lib, in others, a .dll.

Reply:
@elrond79 
Shared libraries(DLLs) on Windows are different than Linux or MacOSX and it is totally normal to get confused by it. 

In a nutshell, when you build a piece of code as a shared library, two files get generated:

1- *.dll 
2- *.lib ( a.k.a import library )

Now, let's say I've decided to use your shared library in my App.  In order to use that, I must tell the linker to link against your *.lib and not *.dll in my project so I can start using your codes. 

let's say, I finished developing my App and want to ship my executable. Since I have a dependency to your shared library, I must  bundle up your shared library with my App executable. In this case, I only need to ship my app with your *.dll.

Unlike MACOSX or LINUX, there is no "run time path" on Windows. Therefore, your *.dll needs to either sit besides the executable or the dlls locations needs to put in the PATH in order to be found by the executable. 

When google test library is built as a shared library on Windows, both *.dll and *.lib get generated for it. *.dll end up in the directory called "bin" and *.lib end up in the directory called "lib".

When you use the ${GTEST_LIBRARIES} variable provided by FindGtest Module, it returns "gtest.lib" which the linker use to link against. This variable is passed to target_link_libraries for the targets that needs it:

target_link_libraries(${TARGET_NAME}
  PRIVATE
      ${GTEST_LIBRARIES}
      ......
)

e.g
al/mayautils/AL/maya/tests/mayaplugintest/CMakeLists.txt
AL/usd/schemas/maya/tests/CMakeLists.txt
....

Since these targets have a dependency to gtest shared library, therefore they need to find the dll when they executed. Therefore, we bundle up the "gtest.dll" which we put in "lib/gtest" so it can be found when the executable runs and we set the the path to the dll location for these targets: "${CMAKE_INSTALL_PREFIX}/lib/gtest"

Here are some good resources if you are really interested in finding more details:

**1-**
CppCon 2017: James McNellis “Everything You Ever Wanted to Know about DLLs”
https://www.youtube.com/watch?v=JPQWQfDhICA

**2-**
https://www.learncpp.com/cpp-tutorial/a1-static-and-dynamic-libraries/

**3-**
https://docs.microsoft.com/en-us/cpp/build/linking-an-executable-to-a-dll?view=vs-2019

Reply:
> When you use the ${GTEST_LIBRARIES} variable provided by FindGtest Module, it returns "gtest.lib" which the linker use to link against. This variable is passed to target_link_libraries for the targets that needs it:
> 
> target_link_libraries(${TARGET_NAME}
> PRIVATE
> ${GTEST_LIBRARIES}

Right - the point I was trying to make was that, if we're doing a release build, then `${GTEST_LIBRARIES}` will end up containing a .lib... which should then be passed to `target_link_libraries`, which, as I understand, is good.

However - if we'd doing a debug build, and `${GTEST_LIBRARY_DEBUG}` is initially `NOTFOUND`, then `${GTEST_LIBRARIES}` will end up pointing at a .dll - which is then passed to `target_link_libraries`, which, as I understand, is bad.

Does that make sense?

Reply:
Partially. To be clear, ${GTEST_LIBRARIES} always return `.lib` On Windows. 

We don't also want to pass around yet another variable ( ${GTEST_LIBRARY_DEBUG} ) for debug.

set(GTEST_LIBRARY ${GTEST_LIBRARY_DEBUG})
set(GTEST_LIBRARIES ${GTEST_LIBRARY})

Both GTEST_LIBRARY and GTEST_LIBRARIES are set accordingly in Debug.

Does any of this make sense? Borrowing your phrase, things should work at this point, I don't want to let the perfect be the enemy of the good. :)

Reply:
> Partially. To be clear, ${GTEST_LIBRARIES} always return `.lib` On Windows.

Assuming we're in debug mode on windows (and the debug lib is not found), we'll hit this line:

https://github.com/Autodesk/maya-usd/blob/b34dc94655c33c791f360a8090a5daedc8eaee97/cmake/Googletest.cmake#L64

...which will set `gtest_library` to have a .dll.

Then when we hit here:

https://github.com/Autodesk/maya-usd/blob/b34dc94655c33c791f360a8090a5daedc8eaee97/cmake/Googletest.cmake#L74

...which makes `GTEST_LIBRARY_DEBUG` have a .dll... 

Then these two lines:

https://github.com/Autodesk/maya-usd/blob/b34dc94655c33c791f360a8090a5daedc8eaee97/cmake/Googletest.cmake#L78
https://github.com/Autodesk/maya-usd/blob/b34dc94655c33c791f360a8090a5daedc8eaee97/cmake/Googletest.cmake#L79

...will make the .dll carry into `GTEST_LIBRARY`, and then `GTEST_LIBRARIES` - so the net result is that `GTEST_LIBRARIES` will have a .dll, NOT a .lib

Or am I missing something obvious?

Reply:
@elrond79 Hmmmmmmmmmmm. It does make sense now that I look closer. I am blind :) My bad!

I went back in history to see why I added this 
` if (GTEST_LIBRARY_DEBUG MATCHES GTEST_LIBRARY_DEBUG-NOTFOUND)` in a first place. 
Almost a year ago this turned out to be a bug somewhere in google test repo. However, this issue is no longer there and I can't really reproduce it on my Windows or MacOS machines. I believe it should be the same on Linux as well but it is worth checking it.

The funny part is that this condition never happens on Debug!!! GTEST_LIBRARY_DEBUG does really return `gtestd.lib` in Debug, even though GTEST_FOUND is still reported as False.

At the end of the day, this is the results that I always had in mind and wanted CMake to give me on Windows:

```
Debug

GTEST_INCLUDE_DIRS                      ----> build/Debug/googletest-install/include
GTEST_LIBRARY/GTEST_LIBRARIES           ----> build/Debug/googletest-install/lib/gtestd.lib
GTEST_MAIN_LIBRARY/GTEST_MAIN_LIBRARIES ----> build/Debug/googletest-install/lib/gtest_maind.lib

Release

GTEST_INCLUDE_DIRS                      ----> build/Release/googletest-install/include
GTEST_LIBRARY/GTEST_LIBRARIES           ----> build/Release/googletest-install/lib/gtest.lib
GTEST_MAIN_LIBRARY/GTEST_MAIN_LIBRARIES ----> build/Release/googletest-install/lib/gtest_main.lib
```
I simplified things a bit. See fc366ecaffe22

Reply:
Ok, great - if we branch isn't a problem anymore, then eliminating it does make things a lot cleaner!  Looks good now!